File: data_sources/transparency_reports/xlsx_to_csv.py
#!/usr/bin/env python3
"""Convert Snap DSA XLSX reports to CSV files for DuckDB ingestion."""

import os
from pathlib import Path
from openpyxl import load_workbook

DATA_DIR = Path(__file__).parent / "data"

def xlsx_to_csvs(xlsx_path, output_dir):
    """Convert each sheet of an XLSX to a separate CSV."""
    wb = load_workbook(xlsx_path, read_only=True)
    os.makedirs(output_dir, exist_ok=True)

    for sheet_name in wb.sheetnames:
        ws = wb[sheet_name]
        csv_name = f"{sheet_name}.csv"
        csv_path = output_dir / csv_name
        rows_written = 0

        with open(csv_path, "w") as f:
            for row in ws.iter_rows():
                cells = []
                for cell in row:
                    val = cell.value
                    if val is None:
                        cells.append("")
                    else:
                        s = str(val).replace('"', '""')
                        if "," in s or '"' in s or "\n" in s:
                            s = f'"{s}"'
                        cells.append(s)
                # Skip fully empty rows
                if all(c == "" for c in cells):
                    continue
                f.write(",".join(cells) + "\n")
                rows_written += 1

        print(f"  {sheet_name}: {rows_written} rows -> {csv_path.name}")

    wb.close()

def main():
    xlsx_files = list((DATA_DIR / "eu_dsa").glob("*.xlsx"))
    if not xlsx_files:
        print("No XLSX files found in data/eu_dsa/")
        return

    for xlsx_path in sorted(xlsx_files):
        stem = xlsx_path.stem
        out_dir = DATA_DIR / "eu_dsa_csv" / stem
        print(f"\nConverting {xlsx_path.name}:")
        xlsx_to_csvs(xlsx_path, out_dir)

if __name__ == "__main__":
    main()

File: data_sources/transparency_reports/xlsx_to_csv.ts
#!/usr/bin/env tsx
/**
 * Convert Snap DSA XLSX reports to CSV files for DuckDB ingestion.
 *
 * Replaces the former Python (openpyxl) script with a pure TypeScript version
 * using SheetJS (xlsx). No Python runtime needed.
 *
 * Usage:
 *   npx tsx data_sources/transparency_reports/xlsx_to_csv.ts
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import * as XLSX from "xlsx";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const DATA_DIR = path.join(__dirname, "data");

function escapeCSVCell(value: unknown): string {
  if (value == null) return "";
  const s = String(value);
  if (s.includes(",") || s.includes('"') || s.includes("\n")) {
    return `"${s.replace(/"/g, '""')}"`;
  }
  return s;
}

function xlsxToCsvs(xlsxPath: string, outputDir: string): void {
  const buf = fs.readFileSync(xlsxPath);
  const workbook = XLSX.read(buf);
  fs.mkdirSync(outputDir, { recursive: true });

  for (const sheetName of workbook.SheetNames) {
    const sheet = workbook.Sheets[sheetName];
    if (!sheet) continue;

    const rows: unknown[][] = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: "" });
    const csvPath = path.join(outputDir, `${sheetName}.csv`);
    let rowsWritten = 0;

    const lines: string[] = [];
    for (const row of rows) {
      const cells = (row as unknown[]).map(escapeCSVCell);
      if (cells.every((c) => c === "")) continue;
      lines.push(cells.join(","));
      rowsWritten++;
    }

    fs.writeFileSync(csvPath, lines.join("\n") + "\n");
    console.log(`  ${sheetName}: ${rowsWritten} rows -> ${path.basename(csvPath)}`);
  }
}

function main(): void {
  const euDsaDir = path.join(DATA_DIR, "eu_dsa");
  if (!fs.existsSync(euDsaDir)) {
    console.log("No data/eu_dsa/ directory found.");
    return;
  }

  const xlsxFiles = fs.readdirSync(euDsaDir)
    .filter((f) => f.endsWith(".xlsx"))
    .sort()
    .map((f) => path.join(euDsaDir, f));

  if (xlsxFiles.length === 0) {
    console.log("No XLSX files found in data/eu_dsa/");
    return;
  }

  for (const xlsxPath of xlsxFiles) {
    const stem = path.basename(xlsxPath, ".xlsx");
    const outDir = path.join(DATA_DIR, "eu_dsa_csv", stem);
    console.log(`\nConverting ${path.basename(xlsxPath)}:`);
    xlsxToCsvs(xlsxPath, outDir);
  }
}

main();
